_posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

```diff
@@ -0,0 +1,709 @@
+    c(bce_loss, dice_loss, loss) %<-% train_batch(b)
+    train_bce <- c(train_bce, bce_loss)
+    train_dice <- c(train_dice, dice_loss)
+    train_loss <- c(train_loss, loss)
+  }
+  
+  torch_save(model, paste0("model_", epoch, ".pt"))
+  
+  cat(sprintf("\nEpoch %d, training: loss:%3f, bce: %3f, dice: %3f\n",
+              epoch, mean(train_loss), mean(train_bce), mean(train_dice)))
+  
+  model$eval()
+  valid_bce <- c()
+  valid_dice <- c()
+  valid_loss <- c()
+  
+  i <- 0
+  for (b in enumerate(valid_dl)) {
+    
+    i <<- i + 1
+    c(bce_loss, dice_loss, loss) %<-% valid_batch(b)
+    valid_bce <- c(valid_bce, bce_loss)
+    valid_dice <- c(valid_dice, dice_loss)
+    valid_loss <- c(valid_loss, loss)
+    
+  }
+  
+  cat(sprintf("\nEpoch %d, validation: loss:%3f, bce: %3f, dice: %3f\n",
+              epoch, mean(valid_loss), mean(valid_bce), mean(valid_dice)))
+}
+
+```
+
+    Epoch 1, training: loss:0.304232, bce: 0.148578, dice: 0.667423
+    Epoch 1, validation: loss:0.333961, bce: 0.127171, dice: 0.816471
+
+    Epoch 2, training: loss:0.194665, bce: 0.101973, dice: 0.410945
+    Epoch 2, validation: loss:0.341121, bce: 0.117465, dice: 0.862983
+
+    [...]
+
+    Epoch 19, training: loss:0.073863, bce: 0.038559, dice: 0.156236
+    Epoch 19, validation: loss:0.302878, bce: 0.109721, dice: 0.753577
+
+    Epoch 20, training: loss:0.070621, bce: 0.036578, dice: 0.150055
+    Epoch 20, validation: loss:0.295852, bce: 0.101750, dice: 0.748757
+
+## Evaluation
+
+In this run, it is the final model that performs best on the validation set. Still, we'd like to show how to load a saved model, using `torch_load()` .
+
+Once loaded, put the model into `eval` mode:
+
+```{r}
+saved_model <- torch_load("model_20.pt") 
+
+model <- saved_model
+model$eval()
+
+```
+
+Now, since we don't have a separate test set, we already know the average out-of-sample metrics; but in the end, what we care about are the generated masks. Let's view some, displaying ground truth and MRI scans for comparison.
+
+```{r}
+# without random sampling, we'd mainly see lesion-free patches
+eval_ds <- brainseg_dataset(valid_dir, augmentation_params = NULL, random_sampling = TRUE)
+eval_dl <- dataloader(eval_ds, batch_size = 8)
+
+batch <- eval_dl$.iter()$.next()
+
+par(mfcol = c(3, 8), mar = c(0, 1, 0, 1))
+
+for (i in 1:8) {
+  
+  img <- batch[[1]][i, , , , drop = FALSE]
+  inferred_mask <- model(img$to(device = device))
+  true_mask <- batch[[2]][i, , , , drop = FALSE]$to(device = device)
+  
+  bce <- nnf_binary_cross_entropy(inferred_mask, true_mask)$to(device = "cpu") %>%
+    as.numeric()
+  dc <- calc_dice_loss(inferred_mask, true_mask)$to(device = "cpu") %>% as.numeric()
+  cat(sprintf("\nSample %d, bce: %3f, dice: %3f\n", i, bce, dc))
+  
+
+  inferred_mask <- inferred_mask$to(device = "cpu") %>% as.array() %>% .[1, 1, , ]
+  
+  inferred_mask <- ifelse(inferred_mask > 0.5, 1, 0)
+  
+  img %>% as.array() %>% .[1, 1, ,] %>% as.raster() %>% plot()
```

Comment:
maybe subset first? just because `.[]` can be confusing...

Reply:
yup!